inpage/src/EthereumApi.ts

```diff
@@ -1,43 +1,118 @@
+import z from 'zod';
+
+import JsonRpcError from './JsonRpcError';
 import assert from './helpers/assert';
 import randomId from './helpers/randomId';
+import { WaxStorage } from './WaxStorage';
+
+const emptyParams = z.union([z.tuple([]), z.undefined()]);
+
+const schema = {
+  eth_requestAccounts: {
+    params: emptyParams,
+    output: z.array(z.string()).min(1),
+  },
+  eth_accounts: {
+    params: emptyParams,
+    output: z.array(z.string()),
+  },
+};
+
+type Schema = typeof schema;
+
+type RequestParams<M extends string> = M extends keyof Schema
+  ? undefined extends z.infer<Schema[M]['params']>
+    ? { params?: z.infer<Schema[M]['params']> }
```

Comment:
Not overly familiar with zod, what's the inner ternary doing here? Is it checking if `params` is optional, then making the RequestParams type reflect that?

Reply:
[zod](https://zod.dev/) is a runtime typing library for TypeScript. It's a better version of io-ts that you may have seen previously in bls-wallet. I believe Zod has since become the defacto standard for doing this kind of thing.

Here's a video about zod: https://www.youtube.com/watch?v=9N50YV5NHaE.

> Is it checking if params is optional, then making the RequestParams type reflect that?

Pretty much. The issue is that if `z.infer<Schema[M]['params']>` is `[] | undefined`, then without the ternary thing we'll end up with `{ method: 'eth_accounts', params: [] | undefined }`. That's a problem because the user will have to write `ethereum.request({ method: 'eth_accounts', params: undefined })` instead of just `ethereum.request({ method: 'eth_accounts' })`.

I hope someday we can use more regular syntax for expressing type logic. I imagine something like this would make it a lot clearer:

```ts
generic RequestParams(M: string) {
  if (!(M extends keyof Schema)) {
    return { params?: unknown[] };
  }

  type Params = z.infer<Schema[M]['params'];

  // Allow omitting the `params` field if it can be `undefined`
  if (undefined extends Params) {
    return { params?: Params };
  }

  return { params: Params };
}
```